pkg/dockerregistry/auth/openshift.go

```diff
@@ -0,0 +1,194 @@
+package openshift
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	ctxu "github.com/docker/distribution/context"
+	repoauth "github.com/docker/distribution/registry/auth"
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"golang.org/x/net/context"
+)
+
+func init() {
+	repoauth.Register("openshift", repoauth.InitFunc(newAccessController))
+}
+
+type AccessController struct {
+	OpenShiftApiVersion string
+	RegistryClient      *dockerregistry.RegistryClient
+}
+
+type authChallenge struct {
+	err error
+}
+
+type OpenShiftAccess struct {
+	Namespace   string
+	ImageRepo   string
+	Verb        string
+	BearerToken string
+}
+
+var _ auth.AccessController = &AccessController{}
+var _ auth.Challenge = &authChallenge{}
+
+// Errors used and exported by this package.
+var (
+	ErrTokenRequired          = errors.New("authorization header with basic token required")
+	ErrTokenInvalid           = errors.New("failed to decode basic token")
+	ErrOpenShiftTokenRequired = errors.New("expected openshift bearer token as password for basic token to registry")
+	ErrNamespaceRequired      = errors.New("repository namespace required")
+	ErrOpenShiftAccessDenied  = errors.New("openshift access denied")
+)
+
+func newAccessController(options map[string]interface{}) (auth.AccessController, error) {
+	fmt.Println("Using OpenShift Auth handler")
+
+	var rc dockerregistry.RegistryClient
+	err := rc.SetRegistryConfig()
+	if err != nil {
+		return nil, err
+	}
+	openshiftApiVersion := os.Getenv("OPENSHIFT_API_VERSION")
+	if len(openshiftApiVersion) == 0 {
+		log.Warn("Env 'OPENSHIFT_API_VERSION' not found, defaulting to 'v1beta1'")
+		openshiftApiVersion = "v1beta1"
+	}
+
+	return &AccessController{
+		OpenShiftApiVersion: openshiftApiVersion.(string),
+		RegistryClient:      &rc,
+	}, nil
+}
+
+// Error returns the internal error string for this authChallenge.
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
+}
+
+// Status returns the HTTP Response Status Code for this authChallenge.
+func (ac *authChallenge) Status() int {
```

Comment:
I would remove this method and just return `http.StatusUnauthorized` directly in `ServeHTTP`